core-js/src/main/javascript/cdf/components/ccc/BaseCccComponent.js

```diff
@@ -77,21 +97,109 @@ define([
+      if (this.chartDefinition) {
+        var cd = this.chartDefinition;
+        // exceptions
+        if (fullName.indexOf("area") > -1 || fullName == "stackedArea") {
+          return "areaStacked";
+        } else if (fullName == "metricDot") {
+          return "bubble";
+        } else if (fullName == "metricLine") {
+          return "scatter";
+        } else if (fullName == "normalizedBar") {
+          return "barNormalized";
+        } else if (fullName == "dot" || fullName == "stackedDot") {
+          return "pointAbstract";
+        } else if (fullName == "stackedLine") {
+          return "line";
+        }
+
+        // general cases
+        if (cd.valuesNormalized) {
+          fullName += "Normalized";
+        }
+        if (cd.orientation === "Horizontal") {
+          fullName += cd.orientation;
+        }
+        if (cd.stacked == true || fullName == "area") {
+          fullName += "Stacked";
+        }
+      }
+
+      return fullName;
+    },
+
     render: function(values) {
+      var myself = this;
+      var cd = this.chartDefinition;
+
+      $("#" + myself.htmlObject).append('<div id="' + myself.htmlObject + 'protovis"></div>');
+      myself._preProcessChartDefinition();
+
+      var triggerVizApiRender = false;
+      if(this._vizApiStyles && !_.contains(chartTypesBlackList, this.getCccChartName())) {
+        triggerVizApiRender = true;
+      }
+
+      if(this.dashboard.isValidQueryDefinition(cd)) {
+        if(triggerVizApiRender) {
+          this.beginQuery(cd, _.bind(myself._renderThroughVizApiDefinitions, myself));
+        } else {
+          this.triggerQuery(cd, _.bind(myself._internalRender, myself));
+        }
+      } else {
+        if(!this.valuesArray) {
+          this.valuesArray = values;
+        }
 
-      $("#" + this.htmlObject).append('<div id="' + this.htmlObject + 'protovis"></div>');
+        if(triggerVizApiRender) {
+          this.execute(_.bind(function() {
+            myself._renderThroughVizApiDefinitions(myself.valuesArray);
+          }, myself));
+        } else {
+          this.synchronous(_.bind(function() {
+            myself._internalRender(myself.valuesArray);
+          }, myself));
+        }
+      } 
+    },
 
-      this._preProcessChartDefinition();
+    /**
+     * 
+     * @param values
+     * @private
+     */
+    _renderThroughVizApiDefinitions: function(values) {
+      var myself = this;
+      
+      context.getAsync("pentaho/ccc/visual/"+myself.getCccChartName()).then(function(View) {
+        myself._internalRender.apply(myself, [values, View.type.extensionEffective]);
+      }).then(_.bind(myself.endExec, myself), _.bind(myself.failExec, myself));
+    },
 
-      var o = $.extend({}, this.chartDefinition);
+    /**
+     *
+     */
+    _internalRender: function(values, externalChartDefinition){
+      var o = externalChartDefinition ? $.extend(externalChartDefinition, this.chartDefinition) : this.chartDefinition;
```

Comment:
Still not sure that simply overwriting whatever the user might have specified in `preProcess` and `postFetch` is the way to go.